### PR TITLE
drivers/segger: Add Segger RTT stream support

### PR DESCRIPTION
**Pull request by Gary-Hobson**
## Summary
Read and write Segger RTT data through the stream interface

### CODE REVIEW

File: include/nuttx/segger/rtt.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+extern "C"
+{
+#endif
+
+/****************************************************************************
+* Name: rtt_channel_initialize
+*****************************************************************************/
+
+#ifdef CONFIG_SEGGER_RTT
+void rtt_channel_initialize(int channel, FAR char *name, size_t upsize,
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:72]: move to new patch

File: drivers/drivers_initialize.c [hunk 2]
```diff
@@ -68,6 +69,16 @@ void drivers_initialize(void)
   devnull_register();   /* Standard /dev/null */
 #endif
 
+#if defined(CONFIG_RTT_CHANNEL1)
```
Comment (xiaoxiang781216) [inline drivers/drivers_initialize.c:72]: move to new patch

File: drivers/segger/Kconfig [hunk 3]
```diff
@@ -77,6 +79,50 @@ config SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL
 
 endchoice # SEGGER_RTT_MODE
 
+endif
+
+config RTT_CHANNEL1
```
Comment (xiaoxiang781216) [inline drivers/segger/Kconfig:84]: move to new patch

File: include/nuttx/segger/rtt.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+#include <nuttx/config.h>
+#include <nuttx/streams.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifdef CONFIG_SEGGER_RTT
+#define rtt_write(stream, buf, len) \
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:36]: create a new patch for the general macro wrapper and move to nuttx/streams.h

File: drivers/segger/Kconfig (hunk 3 above)
Comment (Gary-Hobson) [reply drivers/segger/Kconfig]: done

File: include/nuttx/segger/rtt.h (hunk 4 above)
Comment (xiaoxiang781216) [reply include/nuttx/segger/rtt.h:36]: not move yet.

File: drivers/segger/Make.defs [hunk 5]
```diff
@@ -23,6 +23,7 @@ SGDIR      = segger
 
 ifeq ($(CONFIG_SEGGER_RTT),y)
   CSRCS += segger/RTT/RTT/SEGGER_RTT.c
+  CSRCS += segger/rtt_stream.c
```
Comment (xiaoxiang781216) [inline drivers/segger/Make.defs:26]: should we use stream_rtt.c like syslog_rtt.c?

File: include/nuttx/segger/rtt.h [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+#define rtt_write(stream, buf, len)
+#define rtt_putc(stream, ch)
+#define rtt_getc(stream)
+#endif
+
+/****************************************************************************
+ * Type Declarations
+ ****************************************************************************/
+
+struct lib_rttstream_s
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:51]: let's split to lib_rttoutstream_s and lib_rttinstream_s

**Comment by acassis** [comment]
@xiaoxiang781216 I recall you were doing a documentation about using Segger JLink System View with NuttX, do you have some news about it?

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 I recall you were doing a documentation about using Segger JLink System View with NuttX, do you have some news about it?

@anchao could you take a time:)?

File: drivers/segger/rtt_stream.c [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+ *   stream - User allocated, uninitialized instance of struct
+ *            lib_rttinstream_s to be initialized.
+ *   channel - SEGGER RTT channel number
+ *
+ ****************************************************************************/
+
+void lib_rttinstream(FAR struct lib_rttinstream_s *stream, int channel)
+{
+  DEBUGASSERT(stream != NULL);
+  stream->public.get = rttstream_getc;
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:123]: let's implement gets too

File: drivers/segger/rtt_stream.c [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: rttstream_putc
+ ****************************************************************************/
+
+static void rttstream_putc(FAR struct lib_outstream_s *this, int ch)
+{
+  FAR struct lib_rttoutstream_s *stream = this;
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:44]: need cast

File: drivers/segger/rtt_stream.c [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+/****************************************************************************
+ * Name: rttstream_puts
+ ****************************************************************************/
+
+static int rttstream_puts(FAR struct lib_outstream_s *this,
+                          FAR const void *buf, int len)
+{
+  int ret;
+  FAR struct lib_rttoutstream_s *stream = this;
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:56]: need cast

File: drivers/segger/rtt_stream.c [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  ret = SEGGER_RTT_Write(stream->channel, buf, len);
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:57]: merge with the next line

File: drivers/segger/rtt_stream.c [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+ *   Initializes output stream for use with the configured RTT interface.
+ *
+ * Input Parameters:
+ *   stream - User allocated, uninitialized instance of struct
+ *            lib_rttinstream_s to be initialized.
+ *   channel - SEGGER RTT channel number
+ *
+ ****************************************************************************/
+
+void lib_rttinstream(FAR struct lib_rttinstream_s *stream, int channel)
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:120]: need call SEGGER_RTT_AllocDownBuffer when channel isn't equal zero

File: drivers/segger/rtt_stream.c [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+ *   channel - SEGGER RTT channel number
+ *
+ * Returned Value:
+ *   None (User allocated instance initialized).
+ *
+ ****************************************************************************/
+
+void lib_rttoutstream(FAR struct lib_rttoutstream_s *stream, int channel)
+{
+  DEBUGASSERT(stream != NULL);
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:99]: need call SEGGER_RTT_AllocUpBuffer when channel isn't equal zero

**Comment by anchao** [comment]
>> @xiaoxiang781216 I recall you were doing a documentation about using Segger JLink System View with NuttX, do you have some news about it?

> @anchao could you take a time:)?

@acassis @xiaoxiang781216  Sure, let me prepare the relevant documents and provide them later

File: include/nuttx/segger/rtt.h [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+void lib_rttinstream(FAR struct lib_rttinstream_s *stream, int channel,
+                     size_t bufsize);
+
+
+/****************************************************************************
+ * Name: lib_rttinstream_uninit
+ ****************************************************************************/
+
+void lib_rttinstream_uninit(FAR struct lib_rttinstream_s *stream,
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:83]: move before line 71

File: include/nuttx/segger/rtt.h [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+#ifndef __INCLUDE_NUTTX_SEGGER_RTT_H
+#define __INCLUDE_NUTTX_SEGGER_RTT_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:30]: remove the extra line

File: include/nuttx/segger/rtt.h [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+{
+#endif
+
+#ifdef CONFIG_SEGGER_RTT
+
+/****************************************************************************
+* Name: lib_rttoutstream
+*****************************************************************************/
+
+void lib_rttoutstream(FAR struct lib_rttoutstream_s *stream, int channel,
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:68]:
```suggestion
void lib_rttoutstream(FAR struct lib_rttoutstream_s *stream,
                      int channel, size_t bufsize);
```

File: include/nuttx/segger/rtt.h [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+                      size_t bufsize);
+
+/****************************************************************************
+* Name: lib_rttinstream
+*****************************************************************************/
+
+void lib_rttinstream(FAR struct lib_rttinstream_s *stream, int channel,
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:75]:
```suggestion
void lib_rttinstream(FAR struct lib_rttinstream_s *stream,
                     int channel,  size_t bufsize);
```

File: drivers/segger/rtt_stream.c [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+/****************************************************************************
+ * Name: lib_rttoutstream_uninit
+ ****************************************************************************/
+
+void lib_rttoutstream_uninit(FAR struct lib_rttoutstream_s *stream,
+                             int channel)
+{
+  SEGGER_RTT_ConfigUpBuffer(channel, NULL, NULL, 0,
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:189]: need free memory

File: include/nuttx/segger/rtt.h [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+                     size_t bufsize);
+
+
+/****************************************************************************
+ * Name: lib_rttinstream_uninit
+ ****************************************************************************/
+
+void lib_rttinstream_uninit(FAR struct lib_rttinstream_s *stream,
+                            int channel);
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:84]: remove channel argument

File: include/nuttx/segger/rtt.h [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+
+/****************************************************************************
+ * Name: lib_rttoutstream_uninit
+ ****************************************************************************/
+
+void lib_rttoutstream_uninit(FAR struct lib_rttoutstream_s *stream,
+                             int channel);
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:91]: remove channel argument

File: drivers/segger/rtt_stream.c [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+{
+  char name[8] = {0};
+  FAR char *downbuf = (FAR char *)kmm_malloc(bufsize);
+  DEBUGASSERT(downbuf);
+
+  sprintf(name, "rtt%d", channel);
+  SEGGER_RTT_ConfigUpBuffer(channel, name, downbuf,
+                            bufsize, SEGGER_RTT_MODE_NO_BLOCK_SKIP);
+
+  DEBUGASSERT(stream != NULL);
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:159]: remove

File: drivers/segger/rtt_stream.c [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+void lib_rttinstream(FAR struct lib_rttinstream_s *stream, int channel,
+                     size_t bufsize)
+{
+  char name[8] = {0};
+  FAR char *downbuf = (FAR char *)kmm_malloc(bufsize);
+  DEBUGASSERT(downbuf);
+
+  sprintf(name, "rtt%d", channel);
+  SEGGER_RTT_ConfigUpBuffer(channel, name, downbuf,
+                            bufsize, SEGGER_RTT_MODE_NO_BLOCK_SKIP);
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:157]:
```suggestion
                            bufsize, SEGGER_RTT_MODE_DEFAULT);
```

File: drivers/segger/rtt_stream.c [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+ *
+ * Returned Value:
+ *   None (User allocated instance initialized).
+ *
+ ****************************************************************************/
+
+void lib_rttoutstream(FAR struct lib_rttoutstream_s *stream, int channel,
+                      size_t bufsize)
+{
+  char name[8] = {0};
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:119]:
```suggestion
  char name[32];
```

File: drivers/segger/rtt_stream.c [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+  FAR char *upbuf = (FAR char *)kmm_malloc(bufsize);
+  DEBUGASSERT(upbuf);
+
+  sprintf(name, "rtt%d", channel);
+  SEGGER_RTT_ConfigUpBuffer(channel, name, upbuf,
+                            bufsize, SEGGER_RTT_MODE_NO_BLOCK_SKIP);
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:125]:
```suggestion
                            bufsize, SEGGER_RTT_MODE_DEFAULT);
```

File: drivers/segger/rtt_stream.c [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+ *   stream - User allocated, uninitialized instance of struct
+ *            lib_rttinstream_s to be initialized.
+ *   channel - SEGGER RTT channel number
+ *
+ ****************************************************************************/
+
+void lib_rttinstream(FAR struct lib_rttinstream_s *stream, int channel,
+                     size_t bufsize)
+{
+  char name[8] = {0};
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:151]:
```suggestion
  char name[32];
```

File: drivers/segger/rtt_stream.c (hunk 8 above)
Comment (xiaoxiang781216) [reply drivers/segger/rtt_stream.c:45]: not change? please compile the code locally and ensure no error and warning

File: drivers/segger/rtt_stream.c [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+  return ret;
+}
+
+/****************************************************************************
+ * Name: rttstream_getc
+ ****************************************************************************/
+
+static int rttstream_getc(FAR struct lib_instream_s *this)
+{
+  int ch;
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:69]: need handle the big endian and SEGGER_RTT_Read return nothing correctly:
```
  FAR struct lib_rttinstream_s *stream =
                                (FAR struct lib_rttinstream_s *)this;
  signed char ch = -1;

  DEBUGASSERT(stream);
  stream->public.nget += SEGGER_RTT_Read(stream->channel, &ch, 1);
  return ch;
```

File: drivers/segger/rtt_stream.c [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+/****************************************************************************
+ * Name: rttstream_gets
+ ****************************************************************************/
+
+static int rttstream_gets(FAR struct lib_instream_s *this,
+                          FAR void * buffer, int size)
+{
+  int ret;
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:86]: move after line 87

File: drivers/segger/rtt_stream.c [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+ *   stream - User allocated, uninitialized instance of struct
+ *            lib_rttoutstream_s to be initialized.
+ *   channel - SEGGER RTT channel number
+ *
+ * Returned Value:
+ *   None (User allocated instance initialized).
+ *
+ ****************************************************************************/
+
+void lib_rttoutstream(FAR struct lib_rttoutstream_s *stream, int channel,
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:116]:
```suggestion
void lib_rttoutstream(FAR struct lib_rttoutstream_s *stream,
                      int channel, size_t bufsize)
```

File: drivers/segger/rtt_stream.c [hunk 28]
```diff
@@ -0,0 +1,192 @@
...
+ *   Initializes output stream for use with the configured RTT interface.
+ *
+ * Input Parameters:
+ *   stream - User allocated, uninitialized instance of struct
+ *            lib_rttinstream_s to be initialized.
+ *   channel - SEGGER RTT channel number
+ *
+ ****************************************************************************/
+
+void lib_rttinstream(FAR struct lib_rttinstream_s *stream, int channel,
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:148]:
```suggestion
void lib_rttinstream(FAR struct lib_rttoutstream_s *stream,
                     int channel, size_t bufsize)
```

File: drivers/segger/rtt_stream.c [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+/****************************************************************************
+ * Name: lib_rttinstream_uninit
+ ****************************************************************************/
+
+void lib_rttinstream_uninit(FAR struct lib_rttinstream_s *stream,
+                            int channel)
+{
+  SEGGER_RTT_ConfigDownBuffer(channel, NULL, NULL, 0,
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:177]: need free memory

File: drivers/segger/rtt_stream.c [hunk 30]
```diff
@@ -0,0 +1,192 @@
...
+                              SEGGER_RTT_MODE_NO_BLOCK_SKIP);
+  stream->channel = -1;
+}
+
+/****************************************************************************
+ * Name: lib_rttoutstream_uninit
+ ****************************************************************************/
+
+void lib_rttoutstream_uninit(FAR struct lib_rttoutstream_s *stream,
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:186]: move after line 134

**Comment by Gary-Hobson** [comment]
I have changed to initializing the RTT buffer in lib_rttoutstream, but I still feel like this is not a good design.

This can cause some unnecessary troubles:
When calling lib_rttinstream to initialize channel 0, it will cause an internal static buffer leak, which we cannot release or use

We need to get some internal data structures of Segger RTT in the code, its style is not compatible with nuttx

The initialization of the segger RTT channel is a very independent thing that only needs to be done once. There is no need to process it once in different codes, and there is no need for uninitialization

**Comment by xiaoxiang781216** [comment]
> I have changed to initializing the RTT buffer in lib_rttoutstream, but I still feel like this is not a good design.
> 
> This can cause some unnecessary troubles: When calling lib_rttinstream to initialize channel 0, it will cause an internal static buffer leak, which we cannot release or use
> 

Since channel 0 is initialized by RTT internally, you can just skip the stream initialization code for channel 0.

> We need to get some internal data structures of Segger RTT in the code, its style is not compatible with nuttx
> 

you can simply save fields in lib_rttoutstream_s:
```
struct lib_rttoutstream_s
{
  struct lib_outstream_s public;
  char name[32];
  int channel;
  FAR void *buffer;
};
```

> The initialization of the segger RTT channel is a very independent thing that only needs to be done once. There is no need to process it once in different codes, and there is no need for uninitialization

File: drivers/segger/rtt_stream.c [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+  stream->public.nget = 0;
+  stream->channel = channel;
+}
+
+/****************************************************************************
+ * Name: lib_rttoutstream_uninit
+ ****************************************************************************/
+
+void lib_rttoutstream_uninit(FAR struct lib_rttoutstream_s *stream,
+                             int channel)
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:173]: remove channel

File: drivers/segger/rtt_stream.c [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+  kmm_free(UP->pBuffer);
+  SEGGER_RTT_ConfigUpBuffer(channel, NULL, NULL, 0, SEGGER_RTT_MODE_DEFAULT);
+}
+
+/****************************************************************************
+ * Name: lib_rttinstream_uninit
+ ****************************************************************************/
+
+void lib_rttinstream_uninit(FAR struct lib_rttinstream_s *stream,
+                            int channel)
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:191]: ditto

File: include/nuttx/segger/rtt.h [hunk 33]
```diff
@@ -0,0 +1,96 @@
...
+
+#endif
+
+/****************************************************************************
+ * Name: lib_rttinstream_uninit
+ ****************************************************************************/
+
+void lib_rttinstream_uninit(FAR struct lib_rttinstream_s *stream,
+                            int channel);
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:90]: remove

File: drivers/segger/rtt_stream.c [hunk 34]
```diff
@@ -0,0 +1,203 @@
...
+ * Returned Value:
+ *   None (User allocated instance initialized).
+ *
+ ****************************************************************************/
+
+void lib_rttoutstream(FAR struct lib_rttoutstream_s *stream, int channel,
+                      size_t bufsize)
+{
+  FAR char *name = (FAR char *)kmm_malloc(32);
+  FAR char *upbuf = (FAR char *)kmm_malloc(bufsize);
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:121]: let's use the fix allocation as the suggestion

File: drivers/segger/rtt_stream.c [hunk 35]
```diff
@@ -0,0 +1,203 @@
...
+ *   channel - SEGGER RTT channel number
+ *   bufsize - Size of the RTT channel buffer
+ *
+ ****************************************************************************/
+
+void lib_rttinstream(FAR struct lib_rttinstream_s *stream,
+                     int channel, size_t bufsize)
+{
+  FAR char *name = (FAR char *)kmm_malloc(32);
+  FAR char *downbuf = (FAR char *)kmm_malloc(bufsize);
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:154]: ditto

File: include/nuttx/segger/rtt.h [hunk 36]
```diff
@@ -0,0 +1,96 @@
...
+{
+#endif
+
+#ifdef CONFIG_SEGGER_RTT
+
+/****************************************************************************
+* Name: lib_rttoutstream
+*****************************************************************************/
+
+void lib_rttoutstream(FAR struct lib_rttoutstream_s *stream,
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:66]:
```suggestion
void lib_rttoutstream_open(FAR struct lib_rttoutstream_s *stream,
```

File: include/nuttx/segger/rtt.h [hunk 37]
```diff
@@ -0,0 +1,96 @@
...
+*****************************************************************************/
+
+void lib_rttinstream(FAR struct lib_rttinstream_s *stream,
+                     int channel, size_t bufsize);
+
+/****************************************************************************
+ * Name: lib_rttoutstream_uninit
+ ****************************************************************************/
+
+void lib_rttoutstream_uninit(FAR struct lib_rttoutstream_s *stream,
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:80]:
```suggestion
void lib_rttoutstream_close(FAR struct lib_rttoutstream_s *stream,
```

**Comment by Gary-Hobson** [comment]
done~

it feels so much better now

File: include/nuttx/segger/rtt.h [hunk 38]
```diff
@@ -0,0 +1,94 @@
...
+
+/****************************************************************************
+ * Type Declarations
+ ****************************************************************************/
+
+struct lib_rttoutstream_s
+{
+  struct lib_outstream_s public;
+  char name[32];
+  char *buffer;
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:43]: add FAR

File: include/nuttx/segger/rtt.h [hunk 39]
```diff
@@ -0,0 +1,94 @@
...
+  int channel;
+};
+
+struct lib_rttinstream_s
+{
+  struct lib_instream_s public;
+  char name[32];
+  char *buffer;
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:51]: add FAR

File: include/nuttx/segger/rtt.h [hunk 40]
```diff
@@ -0,0 +1,94 @@
...
+*****************************************************************************/
+
+void lib_rttinstream_open(FAR struct lib_rttinstream_s *stream,
+                          int channel, size_t size);
+
+/****************************************************************************
+ * Name: lib_rttoutstream_close
+ ****************************************************************************/
+
+void lib_rttoutstream_close(FAR struct lib_rttoutstream_s *stream);
```
Comment (xiaoxiang781216) [inline include/nuttx/segger/rtt.h:82]: move before line 70

File: drivers/segger/rtt_stream.c [hunk 41]
```diff
@@ -0,0 +1,188 @@
...
+  stream->channel = channel;
+}
+
+/****************************************************************************
+ * Name: lib_rttoutstream_uninit
+ ****************************************************************************/
+
+void lib_rttoutstream_close(FAR struct lib_rttoutstream_s *stream)
+{
+  kmm_free(stream->buffer);
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:174]: add if ((stream->channel)

File: drivers/segger/rtt_stream.c [hunk 42]
```diff
@@ -0,0 +1,188 @@
...
+  stream->public.gets = rttstream_gets;
+  stream->public.nget = 0;
+  stream->channel = channel;
+}
+
+/****************************************************************************
+ * Name: lib_rttoutstream_uninit
+ ****************************************************************************/
+
+void lib_rttoutstream_close(FAR struct lib_rttoutstream_s *stream)
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:172]: move before line 136

File: drivers/segger/rtt_stream.c [hunk 43]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+  stream->public.get = rttstream_getc;
+  stream->public.gets = rttstream_gets;
+  stream->public.nget = 0;
+  stream->channel = channel;
+}
+
+/****************************************************************************
+ * Name: lib_rttoutstream_uninit
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:169]:
```suggestion
 * Name: lib_rttoutstream_close
```

File: drivers/segger/rtt_stream.c [hunk 44]
```diff
@@ -0,0 +1,188 @@
...
+
+void lib_rttoutstream_close(FAR struct lib_rttoutstream_s *stream)
+{
+  kmm_free(stream->buffer);
+  SEGGER_RTT_ConfigUpBuffer(stream->channel, NULL, NULL,
+                            0, SEGGER_RTT_MODE_DEFAULT);
+}
+
+/****************************************************************************
+ * Name: lib_rttinstream_uninit
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:180]:
```suggestion
 * Name: lib_rttinstream_close
```

File: drivers/segger/rtt_stream.c [hunk 45]
```diff
@@ -0,0 +1,188 @@
...
+ ****************************************************************************/
+
+void lib_rttinstream_close(FAR struct lib_rttinstream_s *stream)
+{
+  kmm_free(stream->buffer);
```
Comment (xiaoxiang781216) [inline drivers/segger/rtt_stream.c:185]: check channel != 0

File: drivers/segger/Make.defs [hunk 46]
```diff
@@ -54,6 +54,10 @@ TARGET_ZIP += $(SGDIR)/RTT.zip
 
 endif
 
+ifeq ($(CONFIG_SEGGER_RTT),y)
+  CSRCS += segger/rtt_stream.c
```
Comment (xiaoxiang781216) [inline drivers/segger/Make.defs:58]:
```suggestion
  CSRCS += segger/stream_rtt.c
```

File: drivers/segger/stream_rtt.c [hunk 47]
```diff
@@ -0,0 +1,192 @@
...
+void lib_rttoutstream_open(FAR struct lib_rttoutstream_s *stream,
+                           int channel, size_t bufsize)
+{
+  if (channel)
+    {
+      stream->buffer = (FAR char *)kmm_malloc(bufsize);
+      DEBUGASSERT(stream->buffer);
+      sprintf(stream->name, "rtt%d", channel);
+      SEGGER_RTT_ConfigUpBuffer(channel, stream->name, stream->buffer,
+                            bufsize, SEGGER_RTT_MODE_DEFAULT);
```
Comment (xiaoxiang781216) [inline drivers/segger/stream_rtt.c:126]:
```suggestion
                               bufsize, SEGGER_RTT_MODE_DEFAULT);
```

File: drivers/segger/stream_rtt.c [hunk 48]
```diff
@@ -0,0 +1,192 @@
...
+  stream->channel = channel;
+}
+
+/****************************************************************************
+ * Name: lib_rttoutstream_close
+ ****************************************************************************/
+
+void lib_rttoutstream_close(FAR struct lib_rttoutstream_s *stream)
+{
+  kmm_free(stream->buffer);
```
Comment (xiaoxiang781216) [inline drivers/segger/stream_rtt.c:142]:
```suggestion
  if (stream->channel)
    {
      kmm_free(stream->buffer);
```

File: drivers/segger/stream_rtt.c [hunk 49]
```diff
@@ -0,0 +1,192 @@
...
+ ****************************************************************************/
+
+void lib_rttinstream_close(FAR struct lib_rttinstream_s *stream)
+{
+  if (stream->channel)
+    {
+      kmm_free(stream->buffer);
+    }
+
+  SEGGER_RTT_ConfigDownBuffer(stream->channel, NULL, NULL,
```
Comment (xiaoxiang781216) [inline drivers/segger/stream_rtt.c:190]: move after line 187

File: drivers/segger/stream_rtt.c [hunk 50]
```diff
@@ -0,0 +1,192 @@
...
+ *   channel - SEGGER RTT channel number
+ *   bufsize - Size of the RTT buffer
+ *
+ * Returned Value:
+ *   None (User allocated instance initialized).
+ *
+ ****************************************************************************/
+
+void lib_rttoutstream_open(FAR struct lib_rttoutstream_s *stream,
+                           int channel, size_t bufsize)
```
Comment (xiaoxiang781216) [inline drivers/segger/stream_rtt.c:118]: should we use SEGGER_BUFFER_SIZE_UP directly?

File: drivers/segger/stream_rtt.c [hunk 51]
```diff
@@ -0,0 +1,192 @@
...
+ * Input Parameters:
+ *   stream - User allocated, uninitialized instance of struct
+ *            lib_rttinstream_s to be initialized.
+ *   channel - SEGGER RTT channel number
+ *   bufsize - Size of the RTT input buffer
+ *
+ ****************************************************************************/
+
+void lib_rttinstream_open(FAR struct lib_rttinstream_s *stream,
+                          int channel, size_t bufsize)
```
Comment (xiaoxiang781216) [inline drivers/segger/stream_rtt.c:162]: should we use BUFFER_SIZE_DOWN?

File: drivers/segger/stream_rtt.c (hunk 50 above)
Comment (xiaoxiang781216) [reply drivers/segger/stream_rtt.c:118]: or add below before line 122:
bufsize = bufsize ? bufsize : SEGGER_BUFFER_SIZE_UP;

File: drivers/segger/stream_rtt.c (hunk 51 above)
Comment (xiaoxiang781216) [reply drivers/segger/stream_rtt.c:165]: or add below before line 169:
bufsize = bufsize ? bufsize : SEGGER_BUFFER_SIZE_DOWN;

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@Gary-Hobson please rebase the patch to the latest master, ci broken is fixed by: https://github.com/apache/nuttx/pull/7706

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**